### PR TITLE
Improved release process

### LINKED ISSUE #13: Publishing does not block when version already exists
The github release action will overwrite the assets of the release but the tag created will point to the initial release commit.  This is not ideal :(

GUS-W-13658882.

### PR DESCRIPTION
**Pull request by colincasey**
This release process supports the following for projects using shared workflows:

- Development and testing of automation changes from the `main` branch 
- Stable, deterministic builds from `v{major}.{minor}.{patch}` refs

For development releases:

- `.github/actions/install-languages-cli/action.yml` must contain an empty value as the default for `download_url` 
> This will trigger the CLI application to be downloaded and installed via Cargo in shared workflows so that the latest changes to Rust code will be used
- Internal actions within shared workflows are locked to the `main` ref so that latest changes to automation will be used
  
For stable releases:
- The previous release version will be bumped according to which option was chosen (`major`, `minor`, or `patch`) when triggering the release workflow
- The Cargo version will be set to this new version
- Internal actions within shared workflows will be updated to use the new version as their ref
- `.github/actions/install-languages-cli/action.yml` will be populated with the location of the CLI asset uploaded to the release as the default value for `download_url` 
> This will trigger the CLI application to be downloaded and installed from a versioned release in shared workflows
- These changes will be committed to a release tag and a new GitHub Release will be created from that

Fixes #13, GUS-W-13658882

### CODE REVIEW

File: .github/workflows/_buildpacks-prepare-release.yml
Comment (edmorley) [inline .github/workflows/_buildpacks-prepare-release.yml:65]: Would something like this work, to avoid the need to generate the args list? 

This makes use of the `format()` function:
https://docs.github.com/en/actions/learn-github-actions/expressions#format

...plus their ternery-like feature:
https://docs.github.com/en/actions/learn-github-actions/expressions#example

```suggestion
        run: actions prepare-release --bump "${{ inputs.bump }}" --repository-url "https://github.com/${{ github.repository }}" ${{ inputs.declarations_starting_version && format('--declarations-starting-version {0}', inputs.declarations_starting_version) }}
```

File: .github/workflows/_buildpacks-prepare-release.yml
Comment (edmorley) [inline .github/workflows/_buildpacks-prepare-release.yml:48]: The repo is public, so I would have thought the token isn't needed? Or is this as a way to test the token works early, rather than only find out right at the end when the "open a PR" step happens? (Or in the case of dry-run mode, the open a PR step doesn't happen at all, so wouldn't be tested.)

**Review by edmorley — COMMENTED** [review]
This looks great so far! 

I've left some initial comments and can have another looks tomorrow :-)

File: .github/workflows/release.yml
Comment (edmorley) [inline .github/workflows/release.yml:88]: The GitHub Actions runners already have `yq` installed:
https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md?plain=1#L104

Could we run it directly, saving the need to rely upon the `mikefarah/yq` action? (This would reduce the supply chain surface area improving security, and mean one less thing that results in Dependabot PR noise.)

File: .github/actions/install-languages-cli/action.yml
Comment (edmorley) [inline .github/actions/install-languages-cli/action.yml:32]: (The `--fail` ensures curl exits non-zero rather than the script trying to run `tar` anyway, and the timeout options prevent hangs)

```suggestion
        curl ${{ inputs.download_url }} --output ${{ runner.temp }}/actions.tar.gz -L --fail --retry 5 --retry-all-errors --connect-timeout 10 --max-time 60
```

(Also I've checked that it's ok to use `` since it was only added in 7.71.0, but it's fine since the GHA 22.04 runner [has v7.81.0-1ubuntu1.13](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#installed-apt-packages))

File: .github/actions/install-languages-cli/action.yml
Comment (edmorley) [inline .github/actions/install-languages-cli/action.yml:2]:
```suggestion
description: Downloads the Languages CLI from a known release or installs using Cargo
```

File: .github/workflows/_buildpacks-release.yml
Comment (colincasey) [reply .github/workflows/_buildpacks-release.yml:241]: it was added back in intentionally.  i realized during testing that the reason i had it set on this step was that the update process requires a published image to exist in order to calculate its digest with `crane` or it fails.  adding the dry run check here means that no modifications will occur and the following PR-related steps will do no work. i'd like to revisit at a later stage though.

File: .github/workflows/_buildpacks-prepare-release.yml
Comment (colincasey) [reply .github/workflows/_buildpacks-prepare-release.yml:65]: hmm, would have to try that to see if it works.

File: .github/workflows/_buildpacks-prepare-release.yml
Comment (colincasey) [reply .github/workflows/_buildpacks-prepare-release.yml:48]: oh, so this is actually related to how you can get signed commits for your GH App.  it's possible to do a bunch of the git setup manually but, if provided a token, the checkout command will automatically configure the git client with the required credentials to use for later commits that can be attributed to the bot.

File: .github/workflows/release.yml
Comment (colincasey) [reply .github/workflows/release.yml:5]: this is also related to getting signed bot commits for the GH App.  if you don't set this then the default committer will be the user who triggered the workflow. 

i like the idea of having it automatically set by https://github.com/heroku/use-app-token-action though. i believe they would have all the information to dynamically fetch this based on the APP_ID we're using.

File: .github/workflows/_buildpacks-release.yml
Comment (edmorley) [reply .github/workflows/_buildpacks-release.yml:241]: Ah that makes sense. Could you add a brief comment mentioning why the `dry-run` is on that step?

File: .github/workflows/release.yml
Comment (colincasey) [reply .github/workflows/release.yml:5]: actually, what might be better is if these were set as organization variables.  that way [heroku/use-app-token-action](https://github.com/heroku/use-app-token-action) would not need to be modified to make extra API calls to retrieve information that is specific to our use case.

File: .github/workflows/_buildpacks-prepare-release.yml
Comment (edmorley) [inline .github/workflows/_buildpacks-prepare-release.yml:48]: I think this new comment is missing the key part - which is answering "why do I need a token for a public repo?". Instead the comment states something that IMO can already be seen/figured out by reading the code (that a token is being set for git auth).

In https://github.com/heroku/languages-github-actions/pull/95#discussion_r1283270681, you mention using the token ensures that git signing is set up as expected - so I think the comment needs to say something roughly like:

> "We set the token here even though it's checking out a public repo, because it ensures XYZ is set up as expected for when we commit changes later." 

...where XYZ is something about git signing.

File: .github/workflows/_buildpacks-release.yml
Comment (edmorley) [inline .github/workflows/_buildpacks-release.yml:244]: (there were a few missing words/typos, and I have shortened slightly)

```suggestion
        # The dry run check is performed here because the update process requires a published
        # image to exist in order to calculate a digest with `crane`. Adding the check here
        # means no files will be modified and so no PR will be created later.
```

File: .github/workflows/release.yml
Comment (edmorley) [inline .github/workflows/release.yml:61]: The `|| echo "0.0.0"` fallback I think adds a failure mode that would result in the older tags being overwritten if the `gh release` command were to fail for say connectivity reasons?

eg imagine:
1. `gh release` fails, even though there is an existing release (eg `1.2.0`), due to network issues or GitHub having a bad day
2. Instead of the workflow aborting/failing, the `|| echo "0.0.0"` fallback means `value` is set to `0.0.0`
3. As such the first bump command ends up being `cargo bump 0.0.0`, meaning the bump following it (eg `cargo bump minor`) would then turn the version into `0.1.0` - even though `0.1.0` is actually older than the current `1.2.0` release.

To avoid this, either:
1. The implementation needs to tell the difference between "there are no existing releases" vs "the gh release command failed for a different reason". (Which is probably more hassle than it's worth to implement)
2. Or, we just remove the `|| echo "0.0.0"` and accept the fact that the implementation here assumes there is an existing release.

Given that we know an existing release exists for this repo, I think (2) might be the easier option?

Also, I think the `2> /dev/null` might (a) be unnecessary (given that pipes and redirects only redirect stdout by default), and (b) might be best removed so if any error occurs it appears in the GitHub Actions log.

File: .github/workflows/release.yml
Comment (edmorley) [inline .github/workflows/release.yml:123]: If the `0.0.0` fallback is removed higher up in the workflow (xref the other thread on this review), then this fallback will need to be removed too.

File: .github/workflows/build-test-release.yml
Comment (edmorley) [inline .github/workflows/build-test-release.yml:63]: With the removal of the old release workflow (plus once #97 is merged), I believe all of the following can be deleted:

```
.github/actions/<everything except the new install cli action>
.github/bootstrap/*
package-lock.json
package.json
tsconfig.json
```

Plus the `node_modules` entry in `.gitignore`.

File: .github/workflows/release.yml
Comment (edmorley) [reply .github/workflows/release.yml:5]: Is the plan to switch to org variables now (and remove these hardcoded lines), or later? (I'm fine with later, it just wasn't clear since this conversation was marked as resolved without context)

Comment (colincasey) [reply .github/workflows/release.yml:5]: sorry, i'll get that sorted out later since it'll require filing a request against another team

File: .github/workflows/release.yml
Comment (colincasey) [reply .github/workflows/release.yml:61]: yes, i was trying to accommodate the "no previous release" edge case but we know there is one already so i'll drop this extra logic.

File: .github/workflows/build-test-release.yml
Comment (colincasey) [reply .github/workflows/build-test-release.yml:63]: done + updated the README to reflect the new setup

File: README.md
Comment (edmorley) [inline README.md:25]: If I click the links when having opened the `README.md` directly (as opposed to viewing when looking at the GitHub overview UI), they currently 404. 

eg Visit https://github.com/heroku/languages-github-actions/blob/06ac54789d81ade15ace09a6e1fa18ae1860b659/README.md and click the links.

I think the only way to avoid that is having to include the full URL sadly.

Also, currently some of the examples are prefixed with `@` but some aren't - I think for clarity/consistency we should prefix them all with `@`.

```suggestion
- the [latest release](https://github.com/heroku/languages-github-actions/releases/latest) version with `@latest`
- a [specific release](https://github.com/heroku/languages-github-actions/releases) version with `@v{major}.{minor}.{patch}`
- the development version with `@main`
```

File: README.md
Comment (edmorley) [inline README.md:86]:
```suggestion
- the [latest release](https://github.com/heroku/languages-github-actions/releases/latest) version with `@latest`
- a [specific release](https://github.com/heroku/languages-github-actions/releases) version with `@v{major}.{minor}.{patch}`
- the development version with `@main`
```

File: README.md
Comment (edmorley) [inline README.md:143]:
```suggestion
- the [latest release](https://github.com/heroku/languages-github-actions/releases/latest) version with `@latest`
- a [specific release](https://github.com/heroku/languages-github-actions/releases) version with `@v{major}.{minor}.{patch}`
- the development version with `@main`
```

File: README.md
Comment (edmorley) [inline README.md:49]: I presume this is `vars.GH_APP_ID` instead of `vars.LINGUIST_GH_APP_ID` (and similar for the secrets section) since you're wanting to keep the README generic, and not tied to our specific Linguist implementation choices? (I'm happy with either approach, so how it is at the moment is fine - I'm only mentioning in case this naming wasn't intentional)

File: README.md
Comment (edmorley) [inline README.md:73]: I think this sentence needs adjusting, since it mentions preparing a release, but this workflow is the "perform a release" one?

File: README.md
Comment (edmorley) [inline README.md:78]: (Feel free to reword this suggestion further - it's still missing an explanation of `[metadata.release]` etc, though maybe that's good enough for now?)

```suggestion
  * Creates an OCI image from the compiled buildpack and publishes it to the Docker Hub repository specified in the buildpack's `buildpack.toml`
```

File: README.md
Comment (edmorley) [inline README.md:79]:
```suggestion
  * Retrieves the OCI image published to Docker Hub and registers this with the CNB Registry
```

File: README.md
Comment (edmorley) [inline README.md:80]:
```suggestion
* Once all buildpacks have been published, updates all the buildpack references found in [heroku/builder](https://github.com/heroku/builder)
```

File: README.md
Comment (edmorley) [inline README.md:131]:
```suggestion
Downloads the Languages CLI from a known release or installs using Cargo.
```

File: README.md
Comment (edmorley) [inline README.md:61]:
```suggestion
| `bump`                          | Which component of the version to increment (major, minor, or patch)                                                                                                                                                            | true     |                             |
```

File: .github/workflows/_buildpacks-prepare-release.yml
Comment (edmorley) [inline .github/workflows/_buildpacks-prepare-release.yml:16]:
```suggestion
        description: Which component of the version to increment (major, minor, or patch)
```

**Review by edmorley — APPROVED** [review]
Amazing thank you! 😄

File: .github/workflows/release.yml
Comment (edmorley) [reply .github/workflows/release.yml:5]: I've filed #99 to track moving these to org variables :-)

File: README.md
Comment (colincasey) [reply README.md:49]: yeah, it was to keep the README generic

File: README.md
Comment (colincasey) [reply README.md]: done

File: README.md
Comment (colincasey) [reply README.md]: added information about the metadata key used

**Pull request merged**
